pkg/storage/stores/shipper/compactor/retention/retention.go

```diff
@@ -0,0 +1,294 @@
+package retention
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/local"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.etcd.io/bbolt"
+
+	"github.com/grafana/loki/pkg/storage"
+	"github.com/grafana/loki/pkg/storage/stores/shipper/util"
+	shipper_util "github.com/grafana/loki/pkg/storage/stores/shipper/util"
+)
+
+var (
+	bucketName  = []byte("index")
+	chunkBucket = []byte("chunks")
+	empty       = []byte("-")
+)
+
+const (
+	logMetricName = "logs"
+	delimiter     = "/"
+	markersFolder = "markers"
+)
+
+type Marker struct {
+	workingDirectory string
+	config           storage.SchemaConfig
+	objectClient     chunk.ObjectClient
+	expiration       ExpirationChecker
+	markerMetrics    *markerMetrics
+}
+
+func NewMarker(workingDirectory string, config storage.SchemaConfig, objectClient chunk.ObjectClient, expiration ExpirationChecker, r prometheus.Registerer) (*Marker, error) {
+	if err := validatePeriods(config); err != nil {
+		return nil, err
+	}
+	metrics := newMarkerMetrics(r)
+	return &Marker{
+		workingDirectory: workingDirectory,
+		config:           config,
+		objectClient:     objectClient,
+		expiration:       expiration,
+		markerMetrics:    metrics,
+	}, nil
+}
+
+func (t *Marker) MarkTableForDelete(ctx context.Context, tableName string) error {
+	start := time.Now()
+	status := statusSuccess
+	defer func() {
+		t.markerMetrics.tableProcessedDurationSeconds.WithLabelValues(tableName, status).Observe(time.Since(start).Seconds())
+		level.Debug(util_log.Logger).Log("msg", "finished to process table", "table", tableName, "duration", time.Since(start))
+	}()
+	level.Debug(util_log.Logger).Log("msg", "starting to process table", "table", tableName)
+
+	if err := t.markTable(ctx, tableName); err != nil {
+		status = statusFailure
+		return err
+	}
+	return nil
+}
+
+func (t *Marker) markTable(ctx context.Context, tableName string) error {
+	objects, err := util.ListDirectory(ctx, tableName, t.objectClient)
+	if err != nil {
+		return err
+	}
+
+	if len(objects) != 1 {
+		// todo(1): in the future we would want to support more tables so that we can apply retention below 1d.
```

Comment:
I think it would be hard to find a window in a large cluster where there is just one file. I think we should run mark phase immediately after compaction which also avoids conflicting updates between compaction and retention process since they run in their independent goroutines and we run compaction in our ops cluster every 5 mins which increases the chance of conflicts.

Reply:
This is just food for thought this comment.

But I thought that may be in the future we want to apply this where compaction is not happening.

Reply:
I would still suggest changing this now because of a couple of reasons:
1. One less way to modify the index.
2. We compact a table only when it has 4 or more files which means retention won't run if we stopped getting writes period and we `>1 count(files) <4`.
3. With a large cluster we run 30+ ingesters and a file could get uploaded as soon as compaction ran which means it would be hard to find a window where there is only one file in an active table.

We can just add a map of tableName->mutex and both compaction and retention code will have to acquire a lock to avoid a race. What do you think?